contracts/components/staking/SlashingController.sol

```diff
@@ -0,0 +1,438 @@
+// SPDX-License-Identifier: UNLICENSED
+// See Forta Network License: https://github.com/forta-protocol/forta-contracts/blob/master/LICENSE.md
+
+pragma solidity ^0.8.4;
+
+import "../BaseComponentUpgradeable.sol";
+import "./SubjectTypes.sol";
+import "./ISlashingExecutor.sol";
+import "./ISlashingController.sol";
+import "./FortaStakingParameters.sol";
+import "../utils/StateMachines.sol";
+import "../../errors/GeneralErrors.sol";
+import "@openzeppelin/contracts/utils/Counters.sol";
+import "@openzeppelin/contracts/utils/math/Math.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "hardhat/console.sol";
+
+contract SlashingController is BaseComponentUpgradeable, ISlashingController, StateMachines, SubjectTypeValidator {
+    using Counters for Counters.Counter;
+
+    enum SlashStates {
+        UNDEFINED,
+        CREATED,
+        REJECTED,
+        DISMISSED,
+        IN_REVIEW,
+        REVIEWED,
+        EXECUTED,
+        REVERTED
+    }
+
+    enum PenaltyMode {
+        UNDEFINED,
+        MIN_STAKE,
+        CURRENT_STAKE
+    }
+
+    struct SlashPenalty {
+        uint256 percentSlashed;
+        PenaltyMode mode;
+    }
+
+    struct Deposit {
+        address proposer;
+        uint256 amount;
+    }
+
+    struct Proposal {
+        uint256 subjectId;
+        address proposer;
+        bytes32 penaltyId;
+        uint8 subjectType;
+    }
+
+    Counters.Counter private _proposalIds;
+    mapping(uint256 => Proposal) public proposals; // proposalId --> Proposal
+    mapping(uint256 => uint256) public deposits; // proposalId --> tokenAmount
+    mapping(bytes32 => SlashPenalty) public penalties; // penaltyId --> SlashPenalty
+    ISlashingExecutor public slashingExecutor;
+    FortaStakingParameters public stakingParameters;
+    IERC20 public depositToken;
```

Comment:
is not planned to be updated, should be marked as immutable.

Reply:
Sorry, what is not planned to be updated? I assume you mean `depositToken` right?

Reply:
yes

Reply:
I was going to say the upgrades plugin warn about that, but we are already doing it in `ForwarderContext`. I assume it is ok since there is no storage slot reserved for this, but why does the plugin warns against this again?

Changed in b2c50a7

Reply:
Yep, it's allowed
https://docs.openzeppelin.com/upgrades-plugins/1.x/faq#why-cant-i-use-immutable-variables